### ISSUE TITLE
nil `semanticTokensProvider` causes error when opening multiple source code files

### ISSUE THREAD

**Issue by jamestrew** [neovim/nvim-lspconfig#2542]
### Description

Setting `client.server_capabilities.sematicTokensProvider = nil` in the `on_attach` throws an error when opening multiple source code files.

```
Error executing vim.schedule lua callback: ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:98: attempt to
index field 'semanticTokensProvider' (a nil value)
stack traceback:
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:98: in function 'tokens_to_ranges'
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:355: in function 'process_response'
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:283: in function 'handler'
        /usr/local/share/nvim/runtime/lua/vim/lsp.lua:1399: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```
This issue originates from upstream but upon further discussion, we suspect that the issue is related to `nvim-lspconfig` and how it calls the `on_attach` function.
https://github.com/neovim/neovim/issues/21588#issuecomment-1493306262

### Neovim version

NVIM v0.9.0-dev-1338+g9e7426718
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3

### Nvim-lspconfig version

0bc0c38e1b11dfb6b5f1794d53869f89ccf9e78f

### Operating system and version

Linux archlinux 6.2.8-arch1-1

### Affected language servers

all with semantic token support

### Steps to reproduce

1. create `minimal_init.lua` from below in your nvim config directory
2. `nvim -nu minimal_init.lua`
3. `:e minimal_init.lua`
4. `:new`
5. `:e init.lua` (some lua file with semantic tokens present)
6. with the focus on `init.lua`, `:LspStop` followed by `:LspStart`

Same steps can be repeated for `tsserver` using two files with `console.log('hello')` in them.
The bug only present itself with code that would be highlighted as a semantic token.

### Actual behavior

Using `lua_ls`, I get the error
```
LSP[lua_ls]: Error SERVER_REQUEST_HANDLER_ERROR: "...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:267: att
empt to index field 'semanticTokensProvider' (a nil value)"
```

With `tsserver`, I get
```
Error executing vim.schedule lua callback: ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:98: attempt to
index field 'semanticTokensProvider' (a nil value)
stack traceback:
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:98: in function 'tokens_to_ranges'
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:355: in function 'process_response'
        ...local/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:283: in function 'handler'
        /usr/local/share/nvim/runtime/lua/vim/lsp.lua:1399: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

### Expected behavior

No error/warnings thrown

### Minimal config

```Lua
-- minimal_init.lua
local on_windows = vim.loop.os_uname().version:match("Windows")

local function join_paths(...)
	local path_sep = on_windows and "\\" or "/"
	local result = table.concat({ ... }, path_sep)
	return result
end

vim.cmd([[set runtimepath=$VIMRUNTIME]])

local temp_dir = vim.loop.os_getenv("TEMP") or "/tmp"

vim.cmd("set packpath=" .. join_paths(temp_dir, "nvim", "site"))

local package_root = join_paths(temp_dir, "nvim", "site", "pack")
local lspconfig_path = join_paths(package_root, "test", "start", "nvim-lspconfig")

if vim.fn.isdirectory(lspconfig_path) ~= 1 then
	vim.fn.system({ "git", "clone", "https://github.com/neovim/nvim-lspconfig", lspconfig_path })
end

vim.cmd.colorscheme("habamax")
vim.lsp.set_log_level("trace")
require("vim.lsp.log").set_format_func(vim.inspect)
local nvim_lsp = require("lspconfig")
local on_attach = function(client, bufnr)
	client.server_capabilities.semanticTokensProvider = nil
	local function buf_set_option(...)
		vim.api.nvim_buf_set_option(bufnr, ...)
	end

	buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")

	-- Mappings.
	local opts = { buffer = bufnr, noremap = true, silent = true }
	vim.keymap.set("n", "gD", vim.lsp.buf.declaration, opts)
	vim.keymap.set("n", "gd", vim.lsp.buf.definition, opts)
	vim.keymap.set("n", "K", vim.lsp.buf.hover, opts)
	vim.keymap.set("n", "gi", vim.lsp.buf.implementation, opts)
	vim.keymap.set("n", "<C-k>", vim.lsp.buf.signature_help, opts)
	vim.keymap.set("n", "<space>wa", vim.lsp.buf.add_workspace_folder, opts)
	vim.keymap.set("n", "<space>wr", vim.lsp.buf.remove_workspace_folder, opts)
	vim.keymap.set("n", "<space>wl", function()
		print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
	end, opts)
	vim.keymap.set("n", "<space>D", vim.lsp.buf.type_definition, opts)
	vim.keymap.set("n", "<space>rn", vim.lsp.buf.rename, opts)
	vim.keymap.set("n", "gr", vim.lsp.buf.references, opts)
	vim.keymap.set("n", "<space>e", vim.diagnostic.open_float, opts)
	vim.keymap.set("n", "[d", vim.diagnostic.goto_prev, opts)
	vim.keymap.set("n", "]d", vim.diagnostic.goto_next, opts)
	vim.keymap.set("n", "<space>q", vim.diagnostic.setloclist, opts)
end

-- Add the server that troubles you here
local name = "lua_ls"
local cmd = { "lua-language-server" } -- needed for elixirls, lua_ls, omnisharp
if not name then
	print("You have not defined a server name, please edit minimal_init.lua")
end
if not nvim_lsp[name].document_config.default_config.cmd and not cmd then
	print([[You have not defined a server default cmd for a server
    that requires it please edit minimal_init.lua]])
end

nvim_lsp[name].setup({
	cmd = cmd,
	on_attach = on_attach,
})

print(
	[[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]
)
```

### LSP log
https://gist.github.com/jamestrew/c64b0aaf3204acaf7cccdf98cff222b7

**Comment by glepnir** [comment]
this is not mini config. we provide a min config please use that.

**Comment by jamestrew** [comment]
Leaving the old min config and steps here just in case. It was VERY reproduce-able last night but now I'm having to alter steps a fair bit. I'm not sure why.

Updating the original post with new steps and correct min config.

<details>
<summary>minimal_init.lua</summary>

```lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
	vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
	vim.fn.system({
		"git",
		"clone",
		"--filter=blob:none",
		"https://github.com/folke/lazy.nvim.git",
		lazypath,
	})
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
	"neovim/nvim-lspconfig",
	config = function()
		require("lspconfig").tsserver.setup({
			on_attach = function(client)
				client.server_capabilities.semanticTokensProvider = nil
			end,
		})
		require("lspconfig").lua_ls.setup({
			on_attach = function(client)
				-- client.server_capabilities.semanticTokensProvider = nil
			end,
		})
	end,
}
require("lazy").setup(plugins, {
	root = root .. "/plugins",
})

vim.cmd.colorscheme("habamax")
vim.lsp.set_log_level('debug') -- log file exists in /tmp/testing/.repro/state/nvim/lsp.log
require('vim.lsp.log').set_format_func(vim.inspect)
```

</details>


```
1. `mkdir -p /tmp/testing`
2. `cd /tmp/testing`
3. create the `minimal_init.lua` and also create a copy `other.lua`
5. `nvim -nu minimal_init.lua .`
6. `:e minimal_init.lua`
7. create a split (`:new`)
8. `:e other.lua`
9. switch to the `minimal_init.lua` split -> very important
10. `:LspStop` followed by `:LspStart`
```

**Comment by glepnir** [comment]
I think issue template had write mini config postion https://github.com/neovim/nvim-lspconfig/blob/master/test/minimal_init.lua .

**Comment by jamestrew** [comment]
I'm sorry I'm not understanding what you mean. I edited the issue description to use the correct mini config which is nearly identical to the one you linked.

**Comment by glepnir** [comment]
I don't use lazy we have truely min config for it why you don't try that?. and i tried your step can't reproduce

![test](https://user-images.githubusercontent.com/41671631/230251999-05373726-d4ff-40ef-8387-1f90b3cf3294.gif)

**Comment by jamestrew** [comment]
I changed the minimal config, steps and lsp log in the issue description after your first comment. 

I'm sorry if I confused you with the original lazy config. Please try to reproduce it again.

Here is a copy:
<details>
<summary>min config</summary>

```lua
-- minimal_init.lua
local on_windows = vim.loop.os_uname().version:match("Windows")

local function join_paths(...)
	local path_sep = on_windows and "\\" or "/"
	local result = table.concat({ ... }, path_sep)
	return result
end

vim.cmd([[set runtimepath=$VIMRUNTIME]])

local temp_dir = vim.loop.os_getenv("TEMP") or "/tmp"

vim.cmd("set packpath=" .. join_paths(temp_dir, "nvim", "site"))

local package_root = join_paths(temp_dir, "nvim", "site", "pack")
local lspconfig_path = join_paths(package_root, "test", "start", "nvim-lspconfig")

if vim.fn.isdirectory(lspconfig_path) ~= 1 then
	vim.fn.system({ "git", "clone", "https://github.com/neovim/nvim-lspconfig", lspconfig_path })
end

vim.cmd.colorscheme("habamax")
vim.lsp.set_log_level("trace")
require("vim.lsp.log").set_format_func(vim.inspect)
local nvim_lsp = require("lspconfig")
local on_attach = function(client, bufnr)
	client.server_capabilities.semanticTokensProvider = nil
	local function buf_set_option(...)
		vim.api.nvim_buf_set_option(bufnr, ...)
	end

	buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")

	-- Mappings.
	local opts = { buffer = bufnr, noremap = true, silent = true }
	vim.keymap.set("n", "gD", vim.lsp.buf.declaration, opts)
	vim.keymap.set("n", "gd", vim.lsp.buf.definition, opts)
	vim.keymap.set("n", "K", vim.lsp.buf.hover, opts)
	vim.keymap.set("n", "gi", vim.lsp.buf.implementation, opts)
	vim.keymap.set("n", "<C-k>", vim.lsp.buf.signature_help, opts)
	vim.keymap.set("n", "<space>wa", vim.lsp.buf.add_workspace_folder, opts)
	vim.keymap.set("n", "<space>wr", vim.lsp.buf.remove_workspace_folder, opts)
	vim.keymap.set("n", "<space>wl", function()
		print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
	end, opts)
	vim.keymap.set("n", "<space>D", vim.lsp.buf.type_definition, opts)
	vim.keymap.set("n", "<space>rn", vim.lsp.buf.rename, opts)
	vim.keymap.set("n", "gr", vim.lsp.buf.references, opts)
	vim.keymap.set("n", "<space>e", vim.diagnostic.open_float, opts)
	vim.keymap.set("n", "[d", vim.diagnostic.goto_prev, opts)
	vim.keymap.set("n", "]d", vim.diagnostic.goto_next, opts)
	vim.keymap.set("n", "<space>q", vim.diagnostic.setloclist, opts)
end

-- Add the server that troubles you here
local name = "lua_ls"
local cmd = { "lua-language-server" } -- needed for elixirls, lua_ls, omnisharp
if not name then
	print("You have not defined a server name, please edit minimal_init.lua")
end
if not nvim_lsp[name].document_config.default_config.cmd and not cmd then
	print([[You have not defined a server default cmd for a server
    that requires it please edit minimal_init.lua]])
end

nvim_lsp[name].setup({
	cmd = cmd,
	on_attach = on_attach,
})

print(
	[[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]
)
```

</details>


1. create `minimal_init.lua` from below in your nvim config directory
2. `nvim -nu minimal_init.lua`
3. `:e minimal_init.lua`
4. `:new`
5. `:e init.lua` (some lua file with semantic tokens present)
6. with the focus on `init.lua`, `:LspStop` followed by `:LspStart`

Log: https://gist.github.com/jamestrew/c64b0aaf3204acaf7cccdf98cff222b7

**Comment by glepnir** [comment]
> create minimal_init.lua from below in your nvim config directory

`nvim --clean  -u  any_dir/.min.lua` is  same. and looks like the step of my gif is same ..

**Comment by wookayin** [comment]
Despite the fix in #2546 (no `vim.schedule` on `on_attach`), I still have this error happening. Indeed `on_attach` is called *AFTER* `textDocument/semanticTokens/full` LSP requests.

In my config, I disable semanticTokens capabilities for `lua_ls` in my own `on_attach` callback.

Here is the reason why and what happens in a chronological order:

1. Open two buffers on vim startup, say `vim -O a.vim b.lua`.

2. [vim.lsp: client._on_attach](https://github.com/neovim/neovim/blob/v0.9.0/runtime/lua/vim/lsp.lua#L1526-L1552)
- L1539: `pcall(config.on_attach, client, bufnr)`
  - This is where lspconfig.on_attach is called.
  - [`lspconfig.on_attach(client, bufnr)`](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/configs.lua#L232-L248)
    - If the `buf` is the current buffer, user's `on_attach` is called.
    - If NOT (e.g. other windows), user's **`on_attach` won't be called immediately**! The execution is deferred until BufEnter.
- L1545-L1548: `semantic_tokens.start(bufnr, client.id)` is called
  - Now that `on_attach` was never called, semantic tokens LSP request will be called because the semanticTokensProvider capabilities were not removed.

3. Now users enter the `lua` buffer, `on_attach` for `lua_ls` called, the semanticTokensProvider set to `nil`.

<details>
   <summary>Stacktrace of `on_attach` for `lua_ls` (executed only after I enter the window) </summary>

```
on_attach lua_ls /Users/wookayin/.dotfiles/nvim/lua/config/lsp.lua
stack traceback:
^I....config/nvim/lua/config/lsp.lua:36: in function '_on_attach'
^I.../nvim-lspconfig/lua/lspconfig/configs.lua:333: in function '_setup_buffer'
^I.../nvim-lspconfig/lua/lspconfig/configs.lua:239: in function <...in/.vim/plugged/nvim-lspconfig/lua/lspconfig/configs.lua:238>
```

Note that the L238 refers to the BufEnter callback.

</details>


4. The asynchronous semanticTokens LSP response arrives (which should not have been fired), and we get the error.

As @bfredl mentioned in https://github.com/neovim/neovim/issues/21588#issuecomment-1493306262, neovim core's lsp on_attach assumes that `on_attach` calls are NOT deferred to the future. However, the actual timing of lspconfig's `on_attach` function is called can be much later, which causes this bug. Shouldn't `on_attach` called immediately?

**Comment by wookayin** [comment]
One workaround is to use `on_init` (or maybe a better place to configure capabilities), though not documented in lspconfig --- but documented in `:help vim.lsp.start_client`.

                   • on_init: Callback (client, initialize_result) invoked
                    after LSP "initialize", where `result` is a table of
                    `capabilities` and anything else the server may send. For
                    example, clangd sends `initialize_result.offsetEncoding`
                    if `capabilities.offsetEncoding` was sent to it. You can
                    only modify the `client.offset_encoding` here before any
                    notifications are sent. Most language servers expect to be
                    sent client specified settings after initialization.
                    Neovim does not make this assumption. A
                    `workspace/didChangeConfiguration` notification should be
                    sent to the server during on_init.


E.g.,
```lua
lspconfig.lua_ls.setup {
  on_init = function(client, initialization_result)
    if client.server_capabilities then
      client.server_capabilities.documentFormattingProvider = false
      client.server_capabilities.semanticTokensProvider = false  -- turn off semantic tokens
    end
  end,
  on_attach = function(client, bufnr)
    -- Do not configure server_capabilities here, as on_attach might be called later
    ...
  end
}
```

which makes the semanticTokens request not sent after the on_attach callback.

**Comment by tomtomjhj** [comment]
@wookayin Hi I think I ran into the same problem. Can you file a new issue for better tracking?

**Comment by wookayin** [comment]
@tomtomjhj I don't think this is a bug worth tracking in a separate issue because I think it's an invalid usage. `semanticTokensProvider` should be done by setting capabilities properly **before** making `lspconfig.setup()` or `vim.lsp.start_client()` call. The configuration done by OP is not appropriate; one should **NOT** tweak `server.capabilities` in `on_attach`. Doing so inside `on_init()` would be fine.
